From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Wed, 31 Jul 2019 20:50:48 +0100
Subject: [PATCH 065/102] userns: add kconfig to set default for unprivileged
 CLONE_NEWUSER

When disabled, unprivileged users will not be able to create
new namespaces. Allowing users to create their own namespaces
has been part of several recent local privilege escalation
exploits, so if you need user namespaces but are
paranoid^Wsecurity-conscious you want to disable this.

By default unprivileged user namespaces are disabled.

Co-authored-by: Jan Alexander Steffens (heftig) <jan.steffens@gmail.com>
Signed-off-by: Jan Alexander Steffens (heftig) <jan.steffens@gmail.com>
Co-authored-by: Levente Polyak <levente@leventepolyak.net>
Signed-off-by: Levente Polyak <levente@leventepolyak.net>
---
 init/Kconfig            | 16 ++++++++++++++++
 kernel/user_namespace.c |  4 ++++
 2 files changed, 20 insertions(+)

diff --git a/init/Kconfig b/init/Kconfig
index 365a853c99c4d..b9d8b5c79796f 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1447,10 +1447,26 @@ config USER_NS
 	  user-space use the memory control groups to limit the amount
 	  of memory a memory unprivileged users can use.
 
 	  If unsure, say N.
 
+config USER_NS_UNPRIVILEGED
+	bool "Allow unprivileged users to create namespaces"
+	depends on USER_NS
+	default n
+	help
+	  When disabled, unprivileged users will not be able to create
+	  new namespaces. Allowing users to create their own namespaces
+	  has been part of several recent local privilege escalation
+	  exploits, so if you need user namespaces but are
+	  paranoid^Wsecurity-conscious you want to disable this.
+
+	  This setting can be overridden at runtime via the
+	  kernel.unprivileged_userns_clone sysctl.
+
+	  If unsure, say N.
+
 config PID_NS
 	bool "PID Namespaces"
 	default y
 	help
 	  Support process id namespaces.  This allows having multiple
diff --git a/kernel/user_namespace.c b/kernel/user_namespace.c
index c84977d660bec..b54a9a25d1c36 100644
--- a/kernel/user_namespace.c
+++ b/kernel/user_namespace.c
@@ -22,11 +22,15 @@
 #include <linux/bsearch.h>
 #include <linux/sort.h>
 #include <linux/nstree.h>
 
 /* sysctl */
+#ifdef CONFIG_USER_NS_UNPRIVILEGED
+int unprivileged_userns_clone = 1;
+#else
 int unprivileged_userns_clone;
+#endif
 
 static struct kmem_cache *user_ns_cachep __ro_after_init;
 static DEFINE_MUTEX(userns_state_mutex);
 
 static bool new_idmap_permitted(const struct file *file,
-- 
2.55.0

